option optimize_for = SPEED;

import "kernel/querydata/idl/querydata_common.proto";
import "kernel/querydata/idl/querydata_structs_client.proto";

package NQueryData;

////////////////////////////////////////
// storage data types
////////////////////////////////////////

/// metadata

message TFactorMeta {
    required string Name = 1;
}

/// data

message TRawFactor {
    required uint32 Id = 1;
    optional bytes StringValue = 2;
    optional int64 IntValue = 3;
    optional float FloatValue = 4;
    optional bytes BinaryValue = 5;
}

message TRawQueryData {
    repeated TRawFactor Factors = 1;
    optional string Json = 2;
    optional string KeyRef = 3;
    optional uint64 Version = 4;
}

enum ETrieVariant {
    TV_COMP_TRIE = 0;
    TV_CODEC_TRIE = 1;
    TV_SOLAR_TRIE = 2;
    TV_METATRIE = 3;
    TV_CODED_BLOB_TRIE = 4;
    TV_CATEG_MASK_COMP_TRIE = 5; // QUERYSEARCH-49
    TV_COUNT = 6;
}

message TFileDescription {
    optional ETrieVariant TrieVariant = 6 [default = TV_COMP_TRIE];
    optional int32 Shards = 8 [default = -1];
    optional int32 ShardNumber = 7 [default = -1];

    required string SourceName = 1;
    optional uint64 Version = 4;
    optional uint64 IndexingTimestamp = 12;

    required EKeyType KeyType = 2; // first subkey
    repeated EKeyType SubkeyTypes = 9; // other subkeys

    repeated TFactorMeta FactorsMeta = 3;
    repeated TFactor CommonFactors = 5;
    optional string CommonJson = 10;
    optional bool HasJson = 11;
    optional bool HasKeyRef = 13;
    optional bool AllowedCommon = 17;

    repeated string IndexerCommandLineOpts = 14;
    optional string IndexerVersion = 15;
    optional string IndexerHostName = 16;
}
